### PR TITLE
Add filtering to property grid

### PR DESCRIPTION
**Pull request by ViliusBa**
Demo on how this looks: 
firefox_cr1jA9JE6m

Note: It is a known issue that filtering does not work correctly with uppercase letters. [An item](https://github.com/iTwin/appui/issues/545) has already been filed that should fix this issue.

### CODE REVIEW

**Review by grigasp — CHANGES_REQUESTED** [review]
Please add an e2e test for this

**Comment by grigasp** [comment]
Another followup item: https://github.com/iTwin/appui/issues/562.

File: packages/itwin/property-grid/src/components/Header.scss [hunk 1]
```diff
@@ -4,12 +4,52 @@
 *--------------------------------------------------------------------------------------------*/
 
 .property-grid-react-panel-header {
-  display: flex;
-  align-items: center;
-  box-sizing: border-box;
-  height: calc(var(--iui-size-l) * 2);
+  display: grid;
+  grid-template-columns: auto auto;
+  grid-template-rows: auto;
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.scss:9]: Maybe it would be possible to reduce css amount by using grid template? Something like this:
```css
.property-grid-react-panel-header {
  display: grid;
  grid-template-areas: "label tools";

  .header-title: {
    grid-area: label;
  }
  .header-tools: {
    grid-area: tools;
  }
  &.search-expanded {
    grid-template-areas: 
      "tools"
      "label";
  }
}
```

File: packages/itwin/property-grid/src/components/SettingsDropdownMenu.tsx [hunk 2]
```diff
@@ -112,7 +112,7 @@ export function SettingsDropdownMenu({ settingsMenuItems, dataProvider }: Settin
   const menuItems = (close: () => void) => settingsMenuItems.map((item, index) => <Fragment key={index}>{item({ dataProvider, close })}</Fragment>);
 
   return <DropdownMenu menuItems={menuItems}>
-    <IconButton styleType="borderless" size="small" title={PropertyGridManager.translate("settings.label")}>
+    <IconButton styleType="borderless" title={PropertyGridManager.translate("settings.label")}>
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/SettingsDropdownMenu.tsx:115]: Have you tried making search bar small? If it looks bad with small search bar I think there are more buttons where you need to remove `size: "small"` (ancestor navigation, more elements button)

**Review by saskliutas — CHANGES_REQUESTED** [review]
Looks like the styling is broken in element list and single element properties grid:
image

image

File: packages/itwin/property-grid/src/components/Header.tsx [hunk 3]
```diff
@@ -6,32 +6,105 @@
...
 
-import type { PropsWithChildren } from "react";
+import { type PropsWithChildren, useEffect, useRef, useState } from "react";
+import type React from "react";
 
 /** @internal */
 export interface HeaderProps {
   className?: string;
   onBackButtonClick?: () => void;
+  titleItem?: React.ReactNode;
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.tsx:19]:
```suggestion
  title?: React.ReactNode;
```

File: packages/itwin/property-grid/src/components/Header.tsx [hunk 4]
```diff
@@ -6,32 +6,105 @@
...
+  headerTools?: React.ReactNode;
+  setSearchInput?: (searchInput: string) => void;
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.tsx:21]:
```suggestion
  onSearchStringChange?: (searchInput: string) => void;
```

File: packages/itwin/property-grid/src/components/Header.tsx [hunk 5]
```diff
@@ -6,32 +6,105 @@
...
+
+  return <SearchBox
+    expandable
+    onCollapse={onClose}
+    onExpand={onOpen}
+    className={className}
+  >
+    <SearchBox.CollapsedState>
+      <SearchBox.ExpandButton
+        title="Expand Searchbar"
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.tsx:59]: Should be localized. And maybe change it to `Open search box`?

File: packages/itwin/property-grid/src/components/Header.tsx [hunk 6]
```diff
@@ -6,32 +6,105 @@
...
+      />
+    </SearchBox.CollapsedState>
+    <SearchBox.ExpandedState >
+      <SearchBox.Input
+        placeholder={"Search"}
+        onChange={(e) => setInputValue(e.currentTarget.value)}
+      />
+      <SearchBox.CollapseButton
+        title="Contract Searchbar"
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.tsx:68]: I think better title for this button would be `Collapse search box` or `Close search box`. Also should be localized.

File: packages/itwin/property-grid/src/components/PropertyGridContent.tsx [hunk 7]
```diff
@@ -127,40 +138,61 @@ interface PropertyGridHeaderProps {
   item?: { className: string, label: PropertyRecord };
   onBackButtonClick?: () => void;
   settingsProps: SettingsDropdownMenuProps;
+  updateSearchText: (searchText: string) => void;
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/PropertyGridContent.tsx:141]:
```suggestion
  onSearchTextChange: (searchText: string) => void;
```

File: packages/itwin/property-grid/src/components/Header.scss [hunk 8]
```diff
@@ -4,12 +4,55 @@
...
+
+        &::before {
+          border: none;
+        }
+      }}
+
+    &.search-bar-expanded {
+      display: grid;
+      grid-template-columns: auto;
+      grid-template-rows: auto auto;
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.scss:43]: I think these can be removed.

File: packages/itwin/property-grid/src/components/Header.scss [hunk 9]
```diff
@@ -4,12 +4,55 @@
...
+
+      .header-title {
+        grid-area: title
+      }
+
+      .header-tools {
+        grid-area: tools;
+        display: flex;
+        align-items: center;
+      }
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.scss:56]: Can be removed. These do not change any styles that are applied when searchbar is collapsed.

File: packages/itwin/property-grid/src/components/Header.scss [hunk 10]
```diff
@@ -4,12 +4,55 @@
...
 
 .property-grid-react-panel-header {
-  display: flex;
-  align-items: center;
-  box-sizing: border-box;
-  height: calc(var(--iui-size-l) * 2);
+  display: grid;
+  grid-template-areas: "title tools";;
+  grid-template-columns: auto auto;
+  grid-template-rows: auto;
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.scss:10]: I think these can be removed

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/itwin/property-grid/src/e2e-tests/PropertyGrid.test.ts-snapshots/property-grid-snapshots-stage-panel-size---narrow-single-element-with-ancestry-navigation-1-chromium-linux.png
Comment (grigasp) [inline packages/itwin/property-grid/src/e2e-tests/PropertyGrid.test.ts-snapshots/property-grid-snapshots-stage-panel-size---narrow-single-element-with-ancestry-navigation-1-chromium-linux.png:1]: This one doesn't look right... I don't think the labels should wrap - instead, we should show an ellipsis at the end.

File: packages/itwin/property-grid/src/components/Header.scss [hunk 11]
```diff
@@ -4,12 +4,59 @@
...
     margin-right: var(--iui-size-xs);
   }
+
+    .property-grid-header-title-text {
+      overflow: hidden;
+      white-space: nowrap;
+      text-overflow: ellipsis;
+    }
+
+    .header-text {
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.scss:19]: I think this should be moved to `PropertyGridContent.scss`.

File: packages/itwin/property-grid/src/components/Header.scss [hunk 12]
```diff
@@ -4,12 +4,59 @@
...
-  box-sizing: border-box;
-  height: calc(var(--iui-size-l) * 2);
-
+  display: grid;
+  grid-template-areas: "title tools";;
   .property-grid-react-header-back-button {
     margin-right: var(--iui-size-xs);
   }
+
+    .property-grid-header-title-text {
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.scss:13]: I think this should be moved to `PropertyGridContent.scss`

File: packages/itwin/property-grid/src/components/Header.tsx [hunk 13]
```diff
@@ -6,32 +6,105 @@
...
+            styleType="borderless"
+            onClick={onBackButtonClick}
+            title={PropertyGridManager.translate("header.back")}
+            className="property-grid-react-header-back-button"
+          >
+            <SvgProgressBackwardCircular />
+          </IconButton>
+          : null
+      }
+      <div className={"header-text"}>
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.tsx:95]: Looks like this `div` is necessary only when using `Header` in `PropertyGridHeader`. I thinks `title` passed from `PropertyGridHeader` can be wrapped into `div` instead of fragment (`<>`).

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/itwin/property-grid/playwright.config.ts
Comment (grigasp) [inline packages/itwin/property-grid/playwright.config.ts:1]: are these changes intentional?

File: packages/itwin/property-grid/src/components/PropertyGridContent.scss [hunk 14]
```diff
@@ -10,8 +10,20 @@
...
+  .property-grid-header-title {
+    width: 100%;
+    overflow: hidden;
+  }
+
+  .property-grid-header-title-text {
+    overflow: hidden;
+    white-space: nowrap;
+    text-overflow: ellipsis;
+  }
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/PropertyGridContent.scss:26]:
```suggestion
  .property-grid-header-title {
    width: 100%;
    overflow: hidden;

    .property-grid-header-title-text {
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    }
  }
```

File: packages/itwin/property-grid/src/components/Header.scss [hunk 15]
```diff
@@ -4,12 +4,48 @@
 *--------------------------------------------------------------------------------------------*/
 
 .property-grid-react-panel-header {
-  display: flex;
-  align-items: center;
-  box-sizing: border-box;
-  height: calc(var(--iui-size-l) * 2);
-
+  display: grid;
+  grid-template-areas: "title tools";;
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.scss:8]:
```suggestion
  grid-template-areas: "title tools";
```

File: packages/itwin/property-grid/src/components/Header.tsx [hunk 16]
```diff
@@ -6,32 +6,103 @@
...
+  onSearchStringChange: (searchInput: string) => void;
+  onClose: () => void;
+  onOpen: () => void;
+  className: string;
+}
+
+function DebouncedSearchBox({ onSearchStringChange, onClose, onOpen, className }: DebouncedSearchBoxProps){
+  const [inputValue, setInputValue] = useState<string>("");
+  const onSearchStringChangeRef = useRef(onSearchStringChange);
+  onSearchStringChangeRef.current = onSearchStringChange;
```
Comment (saskliutas) [inline packages/itwin/property-grid/src/components/Header.tsx:34]: React docs says that you should not change ref value during render. https://react.dev/reference/react/useRef#referencing-a-value-with-a-ref

```suggestion
  useEffect(() => {
    onSearchStringChangeRef.current = onSearchStringChange;
  }, [onSearchStringChange]);
```

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/itwin/property-grid/playwright.config.ts
Comment (ViliusBa) [reply packages/itwin/property-grid/playwright.config.ts:1]: Talked about this in private. Will leave it as it was previously for now.

File: packages/itwin/property-grid/src/components/PropertyGridContent.tsx (hunk 7 above)
Comment (saskliutas) [reply packages/itwin/property-grid/src/components/PropertyGridContent.tsx:141]: Looks like this one is not yet renamed.

File: packages/itwin/property-grid/playwright.config.ts
Comment (grigasp) [reply packages/itwin/property-grid/playwright.config.ts:1]: For tracking, linking the PR to improve e2e running process: https://github.com/iTwin/viewer-components-react/pull/688

File: packages/itwin/property-grid/src/e2e-tests/PropertyGrid.test.ts-snapshots/property-grid-wide-multiple-elements-selected---search-bar-expanded-1-chromium-linux.png
Comment (grigasp) [inline packages/itwin/property-grid/src/e2e-tests/PropertyGrid.test.ts-snapshots/property-grid-wide-multiple-elements-selected---search-bar-expanded-1-chromium-linux.png:1]: This one represents multiple elements' selection. Why are we showing some element's label rather than "Multiple items"?

Comment (ViliusBa) [reply packages/itwin/property-grid/src/e2e-tests/PropertyGrid.test.ts-snapshots/property-grid-wide-multiple-elements-selected---search-bar-expanded-1-chromium-linux.png:1]: fixed

**Review by grigasp — CHANGES_REQUESTED** [review]
Please update the recent widget screenshots to include widget border as in other screenshots we have in `README`. At the moment it doesn't look great without the border:
image
image

**Review by saskliutas — APPROVED** [review]

**Review by grigasp — APPROVED** [review]
